File: src/test/java/net/sf/rails/game/state/PortfolioManagerTest.java
package net.sf.rails.game.state;

import static org.junit.Assert.*;

import net.sf.rails.game.state.Ownable;
import net.sf.rails.game.state.OwnableItem;
import net.sf.rails.game.state.Owner;
import net.sf.rails.game.state.PortfolioManager;
import net.sf.rails.game.state.PortfolioSet;
import net.sf.rails.game.state.Root;

import org.junit.Before;
import org.junit.Test;

public class PortfolioManagerTest {

    private final static String PORTFOLIO_A_ID = "PortfolioA";
    private final static String PORTFOLIO_B_ID = "PortfolioB";
    private final static String OWNER_A_ID = "OwnerA";
    private final static String OWNER_B_ID = "OwnerB";
    private final static String OWNER_C_ID = "OwnerC";
    
    private Root root;
    private PortfolioManager pm;
    private PortfolioSet<Ownable> portfolioA, portfolioB;
    private Owner ownerA, ownerB, ownerC;
    
    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        pm = root.getStateManager().getPortfolioManager();
        ownerA = OwnerImpl.create(root, OWNER_A_ID);
        ownerB = OwnerImpl.create(root, OWNER_B_ID);
        ownerC = OwnerImpl.create(root, OWNER_C_ID);
        portfolioA = PortfolioSet.create(ownerA, PORTFOLIO_A_ID , Ownable.class);
        portfolioB = PortfolioSet.create(ownerB, PORTFOLIO_B_ID , Ownable.class);
        StateTestUtils.close(root);
    }

    @Test
    public void testPMKey() {
        // create various keys
        PortfolioManager.PMKey<?> keyA1, keyA2, keyADiff, keyB;
        keyA1 = pm.createPMKey(Ownable.class, ownerA);
        keyA2 = pm.createPMKey(Ownable.class, ownerA);
        keyADiff = pm.createPMKey(OwnableItem.class, ownerA);
        keyB = pm.createPMKey(Ownable.class, ownerB);

        // check that the two A keys are identical...
        assertTrue(keyA1.equals(keyA2));
        assertTrue(keyA2.equals(keyA1));

        // ... the other differ
        assertFalse(keyA1.equals(keyADiff));
        assertFalse(keyADiff.equals(keyA1));
        assertFalse(keyA1.equals(keyB));
        assertFalse(keyB.equals(keyA1));
        
        // check hashcodes
        assertTrue(keyA1.hashCode() == keyA2.hashCode());
        assertFalse(keyA1.hashCode() == keyADiff.hashCode());
        assertFalse(keyA1.hashCode() == keyB.hashCode());
    }
    
    @Test
    public void testGetUnkownOwner() {
        assertNotNull(pm.getUnkownOwner());
    }

    @Test
    public void testRemovePortfolio() {
        pm.removePortfolio(portfolioA);
        assertNull(pm.getPortfolio(Ownable.class, ownerA));
        
        // undo and redo check
        StateTestUtils.closeAndUndo(root);
        assertSame(portfolioA, pm.getPortfolio(Ownable.class, ownerA));
        StateTestUtils.redo(root);
        assertNull(pm.getPortfolio(Ownable.class, ownerA));
    }

    @Test
    public void testAddPortfolio() {
        // remove first to prepare
        pm.removePortfolio(portfolioA);
        StateTestUtils.close(root);
        assertNull(pm.getPortfolio(Ownable.class, ownerA));
        
        // then add
        pm.addPortfolio(portfolioA);
        assertSame(portfolioA, pm.getPortfolio(Ownable.class, ownerA));
        
        // undo and redo check
        StateTestUtils.closeAndUndo(root);
        assertNull(pm.getPortfolio(Ownable.class, ownerA));

        // redo check
        StateTestUtils.redo(root);
        assertSame(portfolioA, pm.getPortfolio(Ownable.class, ownerA));
    }
    
    @Test
    public void testGetPortfolio() {
        assertSame(portfolioA, pm.getPortfolio(Ownable.class, ownerA));
        assertSame(portfolioB, pm.getPortfolio(Ownable.class, ownerB));
        assertNull(pm.getPortfolio(OwnableItem.class, ownerA));
        assertNull(pm.getPortfolio(Ownable.class, ownerC));
        assertNull(pm.getPortfolio(Ownable.class, pm.getUnkownOwner()));
    }

}

File: dotnet/GameStateTest/PortfolioManagerTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class PortfolioManagerTest
    {
        private const string PORTFOLIO_A_ID = "PortfolioA";
        private const string PORTFOLIO_B_ID = "PortfolioB";
        private const string OWNER_A_ID = "OwnerA";
        private const string OWNER_B_ID = "OwnerB";
        private const string OWNER_C_ID = "OwnerC";

        private Root root;
        private PortfolioManager pm;
        private PortfolioSet<IOwnable> portfolioA, portfolioB;
        private IOwner ownerA, ownerB, ownerC;

        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();
            pm = root.StateManager.PortfolioManager;
            ownerA = OwnerImpl.Create(root, OWNER_A_ID);
            ownerB = OwnerImpl.Create(root, OWNER_B_ID);
            ownerC = OwnerImpl.Create(root, OWNER_C_ID);
            portfolioA = PortfolioSet<IOwnable>.Create(ownerA, PORTFOLIO_A_ID);
            portfolioB = PortfolioSet<IOwnable>.Create(ownerB, PORTFOLIO_B_ID);
            StateTestUtils.Close(root);
        }

        [TestMethod]
        public void TestPMKey()
        {
            // create various keys
            PortfolioManager.PMKey keyA1, keyA2, keyADiff, keyB;
            keyA1 = pm.CreatePMKey<IOwnable>(ownerA);
            keyA2 = pm.CreatePMKey<IOwnable>(ownerA);
            keyADiff = pm.CreatePMKey<OwnableItem<IOwnable>>(ownerA);
            keyB = pm.CreatePMKey<IOwnable>(ownerB);

            // check that the two A keys are identical...
            Assert.IsTrue(keyA1.Equals(keyA2));
            Assert.IsTrue(keyA2.Equals(keyA1));

            // ... the other differ
            Assert.IsFalse(keyA1.Equals(keyADiff));
            Assert.IsFalse(keyADiff.Equals(keyA1));
            Assert.IsFalse(keyA1.Equals(keyB));
            Assert.IsFalse(keyB.Equals(keyA1));

            // check hashcodes
            Assert.IsTrue(keyA1.GetHashCode() == keyA2.GetHashCode());
            Assert.IsFalse(keyA1.GetHashCode() == keyADiff.GetHashCode());
            Assert.IsFalse(keyA1.GetHashCode() == keyB.GetHashCode());
        }

        [TestMethod]
        public void TestGetUnkownOwner()
        {
            Assert.IsNotNull(pm.UnknownOwner);
        }

        [TestMethod]
        public void TestRemovePortfolio()
        {
            pm.RemovePortfolio(portfolioA);
            Assert.IsNull(pm.GetPortfolio<IOwnable>(ownerA));

            // undo and redo check
            StateTestUtils.CloseAndUndo(root);
            Assert.AreSame(portfolioA, pm.GetPortfolio<IOwnable>(ownerA));
            StateTestUtils.Redo(root);
            Assert.IsNull(pm.GetPortfolio<IOwnable>(ownerA));
        }

        [TestMethod]
        public void TestAddPortfolio()
        {
            // remove first to prepare
            pm.RemovePortfolio(portfolioA);
            StateTestUtils.Close(root);
            Assert.IsNull(pm.GetPortfolio<IOwnable>(ownerA));

            // then add
            pm.AddPortfolio(portfolioA);
            Assert.AreSame(portfolioA, pm.GetPortfolio<IOwnable>(ownerA));

            // undo and redo check
            StateTestUtils.CloseAndUndo(root);
            Assert.IsNull(pm.GetPortfolio<IOwnable>(ownerA));

            // redo check
            StateTestUtils.Redo(root);
            Assert.AreSame(portfolioA, pm.GetPortfolio<IOwnable>(ownerA));
        }

        [TestMethod]
        public void TestGetPortfolio()
        {
            Assert.AreSame(portfolioA, pm.GetPortfolio<IOwnable>(ownerA));
            Assert.AreSame(portfolioB, pm.GetPortfolio<IOwnable>(ownerB));
            Assert.IsNull(pm.GetPortfolio<OwnableItem<IOwnable>>(ownerA));
            Assert.IsNull(pm.GetPortfolio<IOwnable>(ownerC));
            Assert.IsNull(pm.GetPortfolio<IOwnable>(pm.UnknownOwner));
        }

    }
}
